### PR TITLE
feat: [CDS-68273]: Added the async validation for template reconcile 

### PR DESCRIPTION
**Pull request by shivamnegi94**
## Describe your changes
Currently for reconciliation, User have to manually click on reconcile button while deploying the pipeline.
There could be scenarios when user make changes to template which is referred to pipeline and forget to reconcile the pipeline before running it which cause pipeline failure.
We have added the reconcile step as part of async validation call.

Is it new task type? No
Did you make changes to the code in a backward-compatible manner? Yes
Did you add a new class which is added in kryo serialization? - No
Will the new version of the delegate receive old version task and process? Yes
If new version of the pipeline-service receives response for the old task will it be able to handle it successfully? Yes
Did you make a change in delegate task which will be created in pipeline-services via grpc call to manager? No change
If the older version of pipeline-service creates the task parameter and calls newer cg manager, will it be okay? Yes
if the newer version of pipeline-service creates the task parameter and calls older cg manager, will it be okay? Yes
Did you remove an enum → No
Did you add an enum(That too at end of list) → No
FeatureFlag Added/Removed -> No
Verify Changes with affects any -> No

Testing:
API:  https://localhost:8181/pipeline/api/pipelines/validate/64520608eac6320db0059def?routingId=kmpySmUISimoRrJL6NL73w&accountIdentifier=kmpySmUISimoRrJL6NL73w&orgIdentifier=default&projectIdentifier=TestProject&pipelineIdentifier=Piprlirtest
Response: `{
    "status": "SUCCESS",
    "data": {
        "status": "SUCCESS",
        "startTs": 1683097095891,
        "endTs": 1683097099448,
        "policyEval": {
            "id": "",
            "deny": false,
            "details": [],
            "message": "FF: [OPA_PIPELINE_GOVERNANCE] is disabled for account: [kmpySmUISimoRrJL6NL73w]",
            "timestamp": "0",
            "status": "",
            "accountId": "",
            "orgId": "",
            "projectId": "",
            "entity": "",
            "type": "",
            "action": "",
            "created": "0"
        },
        "templateValidationResponse": {
            "validYaml": true,
            "exceptionMessage": null
        },
        "validateTemplateInputsResponseDTO": {
            "type": "TemplateInputsErrorMetadataV2",
            "validYaml": false,
            "errorNodeSummary": {
                "nodeInfo": {
                    "identifier": "Piprlirtest",
                    "name": "Piprlirtest",
                    "localFqn": null
                },
                "templateInfo": null,
                "templateResponse": null,
                "childrenErrorNodes": [
                    {
                        "nodeInfo": {
                            "identifier": "Piprlirtest",
                            "name": "Piprlirtest",
                            "localFqn": "pipeline.template"
                        },
                        "templateInfo": {
                            "templateIdentifier": "PipelineTemplateTest",
                            "versionLabel": "v1",
                            "templateEntityType": "Pipeline"
                        },
                        "templateResponse": {
                            "accountId": "kmpySmUISimoRrJL6NL73w",
                            "orgIdentifier": "default",
                            "projectIdentifier": "TestProject",
                            "identifier": "PipelineTemplateTest",
                            "name": "PipelineTemplateTest",
                            "description": "",
                            "tags": {},
                            "yaml": "template:\n  name: PipelineTemplateTest\n  identifier: PipelineTemplateTest\n  versionLabel: v1\n  type: Pipeline\n  projectIdentifier: TestProject\n  orgIdentifier: default\n  tags: {}\n  spec:\n    stages:\n      - stage:\n          name: testTemp\n          identifier: testTemp\n          template:\n            templateRef: StageTemps\n            versionLabel: v2\n            templateInputs:\n              type: Deployment\n              spec:\n                service:\n                  serviceRef: <+input>\n                  serviceInputs: <+input>\n                environment:\n                  environmentRef: <+input>\n                  environmentInputs: <+input>\n                  serviceOverrideInputs: <+input>\n                  infrastructureDefinitions: <+input>\n",
                            "versionLabel": "v1",
                            "templateEntityType": "Pipeline",
                            "templateScope": "project",
                            "version": 0,
                            "gitDetails": {
                                "objectId": null,
                                "branch": null,
                                "repoIdentifier": null,
                                "rootFolder": null,
                                "filePath": null,
                                "repoName": null,
                                "commitId": null,
                                "fileUrl": null,
                                "repoUrl": null,
                                "parentEntityConnectorRef": null,
                                "parentEntityRepoName": null
                            },
                            "entityValidityDetails": {
                                "valid": true,
                                "invalidYaml": null
                            },
                            "lastUpdatedAt": 1683095057067,
                            "storeType": "INLINE",
                            "stableTemplate": true
                        },
                        "childrenErrorNodes": []
                    }
                ]
            },
            "type": "TemplateInputsErrorMetadataV2"
        }
    },
    "metaData": null,
    "correlationId": "5965b1e4-fa6c-4336-9256-5627d6465ab1"
}`

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47480)

### CODE REVIEW

**Comment by shivamnegi94** [comment]
trigger uts

**Review by vaibhav-si — APPROVED** [review]
> Reviewed 1 of 6 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47480)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Comment by shivamnegi94** [comment]
trigger uts

**Review by utkarshchoubey-harness — APPROVED** [review]

File: pipeline-service/service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java [hunk 1]
```diff
@@ -540,11 +545,21 @@ public ResponseDTO<PipelineValidationUUIDResponseBody> startPipelineValidationEv
...
     Optional<PipelineValidationEvent> eventByUuid = pipelineAsyncValidationService.getEventByUuid(uuid);
     PipelineValidationEvent pipelineValidationEvent = eventByUuid.get();
-    PipelineValidationResponseDTO response =
-        PMSPipelineDtoMapper.buildPipelineValidationResponseDTO(pipelineValidationEvent);
+    ValidateTemplateInputsResponseDTO validateTemplateInputsResponseDTO =
+        ValidateTemplateInputsResponseDTO.builder().validYaml(true).build();
+    if (StringUtils.isNotBlank(pipelineIdentifier) && StringUtils.isNotBlank(loadFromCache)) {
+      validateTemplateInputsResponseDTO = pipelineRefreshService.validateTemplateInputsInPipeline(
+          accountId, orgId, projectId, pipelineIdentifier, loadFromCache);
+      if (!validateTemplateInputsResponseDTO.isValidYaml()) {
```
Comment (sribalij) [inline pipeline-service/service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java:556]: This i s already being done inside `pipelineRefreshService.validateTemplateInputsInPipeline`.
Could be a duplicate

**Comment by shivamnegi94** [comment]
trigger uts

**Comment by shivamnegi94** [comment]
trigger uts

**Review by shivamnegi94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47480)* status: 10 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 556 at r4](https://reviewable.io/reviews/harness/harness-core/47480#gh-1191427929:-NVY_-ae6HSSnzKnAap5:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a756f1bdcaa45393b7e0cb59f2fec0ad17184096/pipeline-service/service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L556)):*
<details><summary><i>Previously, sribalij (Sridhar) wrote…</i></summary><blockquote>

This i s already being done inside `pipelineRefreshService.validateTemplateInputsInPipeline`.
Could be a duplicate
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java` line 106 at r7](https://reviewable.io/reviews/harness/harness-core/47480#-NVYX8mV-oXMcUtqaGHg:-NVYZxnx4ZE3xqIp5Ddq:bss4a1x) ([raw file](https://github.com/harness/harness-core/blob/896b8484ea6c14140b6da0a177296c4a28988fc3/pipeline-service/service/src/main/java/io/harness/pms/pipeline/PipelineResourceImpl.java#L106)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why do you need this?
</blockquote></details>

Removed 
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java` line 67 at r7](https://reviewable.io/reviews/harness/harness-core/47480#-NVYXHga7iHLizogSPNm:-NVY_39e852kzWguTWnL:b-hmt8wz) ([raw file](https://github.com/harness/harness-core/blob/896b8484ea6c14140b6da0a177296c4a28988fc3/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java#L67)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

@shivamnegi94 do we need to save - ErrorNodeSummary?
</blockquote></details>

We are using the existing method for validating the template reconcile here.

**Comment by shivamnegi94** [comment]
trigger uts

**Comment by shivamnegi94** [comment]
trigger uts

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47480)* status: 12 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/template/refresh/v2/ValidateTemplateReconcileResponseDTO.java` line 20 at r10](https://reviewable.io/reviews/harness/harness-core/47480#-NViGw_tAUyahZ11MHMF:-NViGw_tAUyahZ11MHMG:b4lk19i) ([raw file](https://github.com/harness/harness-core/blob/03501495e7c866d4e1d9bf0dd950bdeb9517269b/960-ng-core-beans/src/main/java/io/harness/ng/core/template/refresh/v2/ValidateTemplateReconcileResponseDTO.java#L20)):*
> ```Java
> @Builder
> public class ValidateTemplateReconcileResponseDTO {
>   boolean validYaml;
> ```

rename to reconcile
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java` line 67 at r7](https://reviewable.io/reviews/harness/harness-core/47480#-NVYXHga7iHLizogSPNm:-NViH19V0vvcJTKO1dOF:bt7d92e) ([raw file](https://github.com/harness/harness-core/blob/896b8484ea6c14140b6da0a177296c4a28988fc3/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java#L67)):*
<details><summary><i>Previously, utkarshchoubey-harness wrote…</i></summary><blockquote>

we need the errorNodeSummary, as UI uses this to show tree view
</blockquote></details>

done, UI doesnt need in to show red label
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java` line 77 at r10](https://reviewable.io/reviews/harness/harness-core/47480#-NViH7UJ5JquRGjotM0L:-NViH7UJ5JquRGjotM0M:b-3sxq81) ([raw file](https://github.com/harness/harness-core/blob/03501495e7c866d4e1d9bf0dd950bdeb9517269b/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java#L77)):*
> ```Java
>             ValidateTemplateReconcileResponseDTO.builder()
>                 .validYaml(validateTemplateInputsResponseDTO.isValidYaml())
>                 .build());
> ```

We should call after OPA as this should not impact validation status



_Code quote:_
```Java
    ValidationResult validateTemplateInputsValidationResult =
        templateValidationResult.withValidateTemplateReconcileResponseDTO(
            ValidateTemplateReconcileResponseDTO.builder()
                .validYaml(validateTemplateInputsResponseDTO.isValidYaml())
                .build());
```

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47480)* status: 14 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/template/refresh/v2/ValidateTemplateReconcileResponseDTO.java` line 20 at r10](https://reviewable.io/reviews/harness/harness-core/47480#-NViGw_tAUyahZ11MHMF:-NVnv5mHCoG-S3m2gLXm:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/03501495e7c866d4e1d9bf0dd950bdeb9517269b/960-ng-core-beans/src/main/java/io/harness/ng/core/template/refresh/v2/ValidateTemplateReconcileResponseDTO.java#L20)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

rename to reconcile
</blockquote></details>

done
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java` line 77 at r10](https://reviewable.io/reviews/harness/harness-core/47480#-NViH7UJ5JquRGjotM0L:-NVnvmf8CtFQHCuKN0Ke:blrf231) ([raw file](https://github.com/harness/harness-core/blob/03501495e7c866d4e1d9bf0dd950bdeb9517269b/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java#L77)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

We should call after OPA as this should not impact validation status
</blockquote></details>

We should check validationResult status in reconcile if success then only check reconcile else return

**Review by shivamnegi94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47480)* status: 14 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/template/refresh/v2/ValidateTemplateReconcileResponseDTO.java` line 20 at r10](https://reviewable.io/reviews/harness/harness-core/47480#-NViGw_tAUyahZ11MHMF:-NVn7J0hAyPYASQC6pom:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/03501495e7c866d4e1d9bf0dd950bdeb9517269b/960-ng-core-beans/src/main/java/io/harness/ng/core/template/refresh/v2/ValidateTemplateReconcileResponseDTO.java#L20)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

rename to reconcile
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java` line 77 at r10](https://reviewable.io/reviews/harness/harness-core/47480#-NViH7UJ5JquRGjotM0L:-NVnAKdhBQaV0Y1EVzzw:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/03501495e7c866d4e1d9bf0dd950bdeb9517269b/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java#L77)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

We should call after OPA as this should not impact validation status
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/helper/PipelineAsyncValidationHelper.java` line 63 at r12](https://reviewable.io/reviews/harness/harness-core/47480#-NVnvs7j2ZVj_l8pIFYK:-NW0Xq1z41vAduZJWBZL:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8ef06280c41d3c95ce242c5d18ea865208f4bcbd/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/helper/PipelineAsyncValidationHelper.java#L63)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why is this needed now as you are not using it?
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/service/PipelineAsyncValidationServiceImpl.java` line 111 at r12](https://reviewable.io/reviews/harness/harness-core/47480#-NVnvufZ1wXZGCqnDz9H:-NW0Xr3BEQ6bkCAVxYkZ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8ef06280c41d3c95ce242c5d18ea865208f4bcbd/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/service/PipelineAsyncValidationServiceImpl.java#L111)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

not needed
</blockquote></details>

Done.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r8, 3 of 4 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47480)* status: 14 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java` line 127 at r13](https://reviewable.io/reviews/harness/harness-core/47480#-NW0ehIS89Bc93eHhTEp:-NW0ehIS89Bc93eHhTEq:b-2iuhpa) ([raw file](https://github.com/harness/harness-core/blob/3fb1d5f12c009773c6b3f9c40abbc6ad814ce079/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/handler/PipelineAsyncValidationHandler.java#L127)):*
> ```Java
>       ValidationResult validateReconcileResult = governanceValidationResult.withValidateTemplateReconcileResponseDTO(
>           ValidateTemplateReconcileResponseDTO.builder()
>               .reconcileCheck(validateTemplateInputsResponseDTO.isValidYaml())
> ```

This should be reverse condition right? if validYaml then reconcileCheck flag is false, right?



_Code quote:_
```Java
reconcileCheck(validateTemplateInputsResponseDTO.isValidYaml())
```
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/helper/PipelineAsyncValidationHelper.java` line 63 at r12](https://reviewable.io/reviews/harness/harness-core/47480#-NVnvs7j2ZVj_l8pIFYK:-NW0ep_W9wVuvcoyY4oy:bwld33b) ([raw file](https://github.com/harness/harness-core/blob/8ef06280c41d3c95ce242c5d18ea865208f4bcbd/pipeline-service/service/src/main/java/io/harness/pms/pipeline/validation/async/helper/PipelineAsyncValidationHelper.java#L63)):*
<details><summary><i>Previously, shivamnegi94 (Shivam Negi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I dont see this method getting removed

**Comment by shivamnegi94** [comment]
trigger uts

**Comment by shivamnegi94** [comment]
trigger uts

**Review by utkarshchoubey-harness — APPROVED** [review]

**Comment by shivamnegi94** [comment]
trigger uts

**Comment by shivamnegi94** [comment]
trigger review

**Comment by shivamnegi94** [comment]
trigger messagecheck

**Pull request merged**